src/it/java/org/checkstyle/suppressionxpathfilter/AbstractXpathRegressionTest.java

```diff
@@ -0,0 +1,207 @@
+            throws Exception {
+        final FileText fileText = new FileText(fileToProcess,
+                StandardCharsets.UTF_8.name());
+        final DetailAST rootAst = JavaParser.parseFile(fileToProcess,
+                JavaParser.Options.WITH_COMMENTS);
+        final XpathQueryGenerator queryGenerator = new XpathQueryGenerator(rootAst,
+                position.violationLineNumber, position.violationColumnNumber,
+                fileText, DEFAULT_TAB_WIDTH);
+
+        return queryGenerator.generate();
+    }
+
+    private static DefaultConfiguration createCheckerWithTreeWalkerModules(
+            DefaultConfiguration... children) {
+        final DefaultConfiguration treeWalkerConfig = createModuleConfig(TreeWalker.class);
+        for (DefaultConfiguration child : children) {
+            treeWalkerConfig.addChild(child);
+        }
+        return createRootConfig(treeWalkerConfig);
+    }
+
+    private static void verifyXpathQueries(List<String> generatedXpathQueries,
+                                           List<String> expectedXpathQueries) {
+        assertEquals("Generated queries do not match expected ones", expectedXpathQueries,
+                generatedXpathQueries);
+    }
+
+    /**
+     * Returns canonical path for the file with the given check name and file name.
+     * @param checkName check name.
+     * @param filename file name.
+     * @return canonical path for the file name.
+     * @throws IOException if I/O exception occurs while forming the path.
+     */
+    protected String getPath(String checkName, String filename) throws IOException {
+        final String subpackage = checkName.toLowerCase(Locale.getDefault())
+                .replace("check", "");
+        return new File("src/it/resources/" + getPackageLocation()
+                + "/" + subpackage + "/" + filename)
+                .getCanonicalPath();
+    }
+
+    private String createSuppressionsXpathConfigFile(String checkName,
+                                                     List<String> xpathQueries)
+            throws Exception {
+
+        final File suppressionsXpathConfigFile = temporaryFolder.newFile();
+        try (Writer bw = Files.newBufferedWriter(suppressionsXpathConfigFile.toPath(),
+                StandardCharsets.UTF_8)) {
+            bw.write("<?xml version=\"1.0\"?>\n");
+            bw.write("<!DOCTYPE suppressions PUBLIC\n");
+            bw.write("    \"-//Puppy Crawl//DTD Suppressions Xpath Experimental 1.2//EN\"\n");
+            bw.write("    \"http://checkstyle.sourceforge.net/dtds/");
+            bw.write("suppressions_1_2_xpath_experimental.dtd\">\n");
+            bw.write("<suppressions>\n");
+            bw.write("   <suppress-xpath\n");
+            bw.write("       checks=\"");
+            bw.write(checkName);
+            bw.write("\"\n");
+            bw.write("       query=\"");
+            bw.write(xpathQueries.stream().collect(Collectors.joining(DELIMITER)));
```

Comment:
why "join" ? I asked this questions before.
if you postponing it for another issue please reference it or define reason.

Reply:
I probably misunderstood you
There are two ways
1. Create many `SuppressionXpathFilter` modules with only xpath query
2. Create one `SuppressionXpathFilter` with joined xpath query

I see no difference between them, the only difference in the second case there is an additional opportunity to test complex, combined queries

Reply:
Why we need all of queries to suppress violation but not the one one of them ?

Reply:
Because all these queries point to the same line and column number

How do we know which query is responsible for node we want to suppress, since violation message does not contain token type information, only line and column?

So for each test we can manually specify required xpath query, or use one of two approaches I described above

Reply:
> my experiment:

First of all about different column numbers
The reason of this difference, that user messages consider tabs and start with 1, when `AstTreeStringPrinter ` prints 0-based column numbers without calculating tabs.

So If you want numbers to be the same, we should pass tabwidth to `AstTreeStringPrinter` and little change one line and thats all

> So if we force Checks to report issue on AST nodes that have text from file

You are suggesting to change `log()` calls inside checks, so Checks will report violation on nearest node that satisfies condition `node.getText()` not equals string representation of the `node.getType()`.
if you mean it, it will require a lot of correcting tests. But yes, probably it will work

Another way is to print `tokenType` inside violation messages, for instance
`InputTest.java:2:5:METHOD_DEF: Cyclomatic Complexity is 2 (max allowed is 0). [CyclomaticComplexity]`
And make xpath query generator accept tokenType among line and column numbers

Third approach is very similar to [#102](https://github.com/checkstyle/checkstyle/issues/102)
This is more natural for me and requires much less pain to make it
We could force users to use configuration file alongside with query generator parameter
So if you want to generate xpath query will use the following format
`java -jar checkstyle-8.10-all.jar -c checks.xml -s 4:5 TestClass.java`
Since we pass configuration file, we will be able to extract `LocalizedMessage` from errors list during runtime and generate xpath queries

Reply:
Lets move out of this huge PR:
#5877 - AstTreeStringPrinter should use 1-based column and consider tab size
#5878 - Generate single xpath query by CLI